### PR TITLE
Prevent "missing field" error logging for overlapping nested deferred fields. - #13454

### PR DESCRIPTION
**Pull request by phryneas**
Supersedes #13454, based on release-4.3

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed incorrect “Missing field” errors when deferred fragments contain fields selected both inside and outside the fragment.
  - Improved handling of nested and failed `@defer` responses so pending fields are not reported prematurely.
  - Preserved missing-field warnings for fields that are not part of deferred content.

- **API Updates**
  - Added APIs for checking whether deferred content is still pending during incremental requests and cache operations.
  - Added typed support for passing incremental delivery information through cache writes.

### CODE REVIEW

File: src/cache/inmemory/types.ts [hunk 1]
```diff
@@ -136,7 +137,14 @@ export type ReadQueryOptions = {
 /** @internal */
 export interface DiffIncrementalInfo {
   streamInfo?: StreamInfoTrie;
-  deferInfo?: DeferInfoTrie;
```
Comment (phryneas) [inline src/cache/inmemory/types.ts:139]: `deferInfo` being a `Trie` worked for the alpha 19 protocol, but it won't work for defer20220824 where we want to always assume "true" - it would start failing at the point where we can't access array indices. So switching this to a function style signature here, instead of keeping both.

File: src/cache/core/types/Cache.ts [hunk 2]
```diff
@@ -302,6 +303,12 @@ export declare namespace Cache {
     extensions?: ExtensionsWithStreamInfo;
   }
 
+  namespace WriteQueryOptions {
+    interface WithIncremental<TData, TVariables extends OperationVariables>
+      extends WriteQueryOptions<TData, TVariables>,
+        WithHandleIncremental {}
+  }
```
Comment (phryneas) [inline src/cache/core/types/Cache.ts:310]: This needed to be an interface to extend `WriteQueryOptions` - writing `WriteQueryOptions<TData, TVariables> & WithHandleIncremental` would potentially break assumptions around `VariableOptions` in some of our type tests.

Moving it here seemed sensible.

File: src/cache/inmemory/readFromStore.ts [hunk 3]
```diff
@@ -894,14 +892,14 @@ export class StoreReader {
...
-      if (context.deferInfo && isDeferredFragment(selection, variables)) {
+      if (context.isDeferPending && isDeferredFragment(selection, variables)) {
         const directive = selection.directives?.find(
           (d) => d.name.value === "defer"
         );
         const label =
           directive && getDirectiveArgValue(directive, "label", Kind.STRING);
 
-        prune = !!context.deferInfo.peekArray(path.concat(label || []));
+        prune = context.isDeferPending(path, label || undefined);
```
Comment (jerelmiller) [inline src/cache/inmemory/readFromStore.ts]:
```suggestion
        prune = context.isDeferPending(path, label);
```

File: src/cache/inmemory/writeToStore.ts [hunk 4]
```diff
@@ -339,11 +345,16 @@ export class StoreWriter {
...
+      currentPath,
       typename
     ).forEach((context, field) => {
       const resultFieldKey = resultKeyNameFromField(field);
       const value = result[resultFieldKey];
-      const path = [...currentPath, field.name.value];
+      // Use the response key (which accounts for aliases), not
+      // field.name.value, so that @defer/@stream path lookups (and
+      // context.isDeferPending) resolve to the correct boundary for
+      // aliased fields.
```
Comment (jerelmiller) [inline src/cache/inmemory/writeToStore.ts]:
```suggestion
```

File: src/cache/inmemory/writeToStore.ts [hunk 5]
```diff
@@ -339,11 +345,16 @@ export class StoreWriter {
...
+      const path = [...currentPath, resultFieldKey];
```
Comment (jerelmiller) [inline src/cache/inmemory/writeToStore.ts:353]: This change might be worth calling out in the changeset because this also fixes `@stream` fields that have aliases somewhere in the path (you can trace this passed all the way to `policies.runMergeFunction` which uses it for `streamInfo.deref().peekArray(path)`).

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts [hunk 6]
```diff
@@ -9,12 +9,21 @@ import {
...
 import { markAsStreaming, spyOnConsole } from "@apollo/client/testing/internal";
 import { addTypenameToDocument } from "@apollo/client/utilities";
-import type { DeferInfoTrie } from "@apollo/client/utilities/internal";
 import {
   handleIncrementalSymbol,
   makeStreamInfoTrie,
 } from "@apollo/client/utilities/internal";
 
+function pendingDefers(
+  ...entries: Array<[path: Array<string | number>, label?: string]>
```
Comment (jerelmiller) [inline src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: Image

(sorry couldn't resist 🤣)

That said... the 2d array thing makes tests very hard to follow. Can we change these into objects instead?

```suggestion
  ...entries: Array<{ path: Array<string | number>, label?: string }>
```

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts [hunk 7]
```diff
@@ -9,12 +9,21 @@ import {
 } from "@apollo/client/cache";
 import { markAsStreaming, spyOnConsole } from "@apollo/client/testing/internal";
 import { addTypenameToDocument } from "@apollo/client/utilities";
-import type { DeferInfoTrie } from "@apollo/client/utilities/internal";
 import {
   handleIncrementalSymbol,
   makeStreamInfoTrie,
 } from "@apollo/client/utilities/internal";
 
+function pendingDefers(
```
Comment (jerelmiller) [inline src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: Could we maybe call this `getIsDeferPending` or `makeIsDeferPending` since its the helper function to create `isDeferPending`?

```suggestion
function getIsDeferPending(
```
or
```suggestion
function makeIsDeferPending(
```

Comment (jerelmiller) [reply src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: Oh and can we move this function to the bottom of the file with the other helper function?

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts [hunk 8]
```diff
@@ -9742,15 +9744,16 @@ test("does not reuse a deferInfo-pruned result when deferInfo is later omitted",
     },
   });
 
-  const deferInfo: DeferInfoTrie = new Trie();
-  deferInfo.lookup("greeting");
+  // A fresh callback identity per call, just like a fresh Trie was, since the
+  // callback participates in the diff's memoization key.
```
Comment (jerelmiller) [inline src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]:
```suggestion
```

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts [hunk 9]
```diff
@@ -9819,16 +9822,16 @@ test("does not reuse a deferInfo-pruned result when a sibling boundary is delive
   });
 
   {
-    const deferInfo: DeferInfoTrie = new Trie();
-    deferInfo.lookup("greeting");
-    deferInfo.lookup("hero");
+    // A fresh callback identity per call, just like a fresh Trie was, since
+    // the callback participates in the diff's memoization key.
```
Comment (jerelmiller) [inline src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]:
```suggestion
```

File: src/incremental/handlers/defer20220824.ts [hunk 10]
```diff
@@ -120,6 +128,14 @@ class DeferRequest<TData extends Record<string, unknown>>
           }
         }
 
+        if ("data" in incremental && incremental.data === null) {
+          // A `null` `data` on a `@defer` incremental chunk means that
+          // boundary failed and will never deliver its data. We can't tell
+          // which boundary failed, so treat any boundary as potentially
+          // still pending until the whole request finishes.
```
Comment (jerelmiller) [inline src/incremental/handlers/defer20220824.ts]: Correct me if I'm wrong, but I don't think this comment is accurate. 

> still pending until the whole request finishes

This seems wrong because once this is `true` (any chunk fails), then all future checks for this request will report all defer chunks as pending since `isDeferPending` uses `||`.

To be clear, I think the logic below is probably right, but this comment is misleading. That said, do we need the comment at all? This is essentially a repeated comment from `isDeferPending` below, so I would just delete this.


```suggestion
```

File: src/incremental/handlers/defer20220824.ts [hunk 11]
```diff
@@ -120,6 +128,14 @@ class DeferRequest<TData extends Record<string, unknown>>
           }
         }
 
+        if ("data" in incremental && incremental.data === null) {
```
Comment (jerelmiller) [inline src/incremental/handlers/defer20220824.ts]: Do we get a `path` for this condition? Let's double check, because we certainly merge `data: null` into our result at path if thats the case, which still satisfies the completeness of the query. I'm not sure we need this check otherwise.

File: src/incremental/handlers/defer20220824.ts [hunk 12]
```diff
@@ -157,6 +173,12 @@ class DeferRequest<TData extends Record<string, unknown>>
 
     return result;
   }
+
+  // This protocol can't identify individual `@defer`/`@stream` boundaries by
+  // `path`/`label`, so we can't answer whether a *specific* boundary is
+  // pending - only whether *anything* might still be coming (`hasNext`), or
+  // whether a boundary failed and will never arrive (`hasFailedDefer`).
+  isDeferPending = () => this.hasNext || this.hasFailedDefer;
```
Comment (jerelmiller) [inline src/incremental/handlers/defer20220824.ts]: We could make this a little bit smarter by at least detecting if the query has any `@defer` directives in it. `startRequest` gets the `query` used for this request, its not just passed to this constructor because we decided to hold off using it until we needed it. I think this is probably a good enough reason to do so now.

If we update this class constructor to accept the `query`, then we can check if `@defer` is part of it. If its a purely `@stream` query, we can return `false` here and allow the query to continue warning.

File: src/incremental/handlers/graphql17Alpha9.ts [hunk 13]
```diff
@@ -124,6 +125,24 @@ class IncrementalRequest<TData>
...
+      if (pending.id in this.streamPositions) {
+        continue;
+      }
+
+      if (pending.label === label && equal(pending.path, path)) {
+        return this.completedMap.get(pending.id) !== true;
+      }
+    }
+
+    return false;
```
Comment (jerelmiller) [inline src/incremental/handlers/graphql17Alpha9.ts:148]: If we're going to loop again, I'd prefer that we use `getPendingWithInfo` instead which records what type the pending entry is and whether its been delivered:

```suggestion
    for (const pending of this.getPendingWithInfo()) {
      if (
        pending.type === "defer" && 
        equal(pending.path, path) && 
        pending.label === label
      ) {
        return !pending.delivered;
      }
    }

    return false;
```

That said, its a bit of a smell to me that we have 2 methods that essentially do the same thing. I'm going to experiment a bit more to see if we can reduce this down to a single method (or split them up at least).

File: src/core/QueryInfo.ts [hunk 14]
```diff
@@ -312,10 +314,19 @@ export class QueryInfo<
...
+          // `this.incremental.isDeferPending` itself. `context.isDeferPending`
+          // is part of the memoization key of
+          // `prunePartialBoundaries`/`prunePartialStreamArray` in
+          // readFromStore.ts. A stable function reference would let a prune
+          // result computed while a boundary was pending be served again
+          // after it was delivered. A fresh identity per read (like the
+          // fresh Trie it replaces) keeps those results from being reused
+          // across chunks, while `undefined` (nothing pending) still allows
+          // reuse. Do not "optimize" this into a cached reference.
+          const isDeferPending = this.incremental!.isDeferPending;
```
Comment (jerelmiller) [inline src/core/QueryInfo.ts:327]: We really don't need this in the loop since this now only executes once. We needed this before because we needed to call `lookupArray` for every `defer` entry's `path`, but now that its a single function, this can be removed.

That really only leaves the `else if` path, which makes me wonder if `getPendingWithInfo` is useful anymore. As mentioned in my comment for the handler, I think we should consider splitting that method apart since the `defer` "branch" of `getPendingWithInfo` is more-or-less dead.

File: src/incremental/handlers/graphql17Alpha9.ts [hunk 15]
```diff
@@ -124,6 +125,24 @@ class IncrementalRequest<TData>
     });
   }
 
+  /** @internal */
+  isDeferPending = (
```
Comment (jerelmiller) [inline src/incremental/handlers/graphql17Alpha9.ts:134]: It might be worth a comment here explaining why this needs to be `isDeferPending =` rather than `isDeferPending() {}` since we need a new instance method for each request to avoid cache keys issues for the prune* methods in `readFromStore`. Its easy to miss that in the future since this feels out of place next to the rest of the methods in this class.

File: src/incremental/handlers/defer20220824.ts [hunk 16]
```diff
@@ -65,6 +65,14 @@ class DeferRequest<TData extends Record<string, unknown>>
 {
   public hasNext = true;
 
+  // This protocol has no `pending`/`completed` bookkeeping, so we can't tell
+  // which `@defer` boundary (if any) a given `path`/`label` refers to - only
+  // whether *something* is still in flight (`hasNext`) or a boundary failed
+  // outright and will never deliver (`hasFailedDefer`). `isDeferPending`
+  // below is intentionally coarse-grained: it treats every path as pending
+  // as long as either is true.
```
Comment (jerelmiller) [inline src/incremental/handlers/defer20220824.ts]:
```suggestion
```

This is essentially a repeat of the other 2 comments. I'd just use the one at `isDeferPending` if anything, so let's delete it.

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts [hunk 17]
```diff
@@ -10857,3 +10857,293 @@ test.failing(
...
+    mockDeferStreamGraphQL17Alpha9();
+
+  const client = new ApolloClient({
+    link: httpLink,
+    cache: new InMemoryCache(),
+    incrementalHandler: new GraphQL17Alpha9Handler(),
+  });
+  const stream = new ObservableStream(client.watchQuery({ query }));
+
+  // `text` and `name` are missing, but not deferred
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:11074]: Could we avoid setting up the test in this way? Yes this exercises the "missing field" situation, but it does so under the premise of the server sending these values so this is essentially mocking a misbehaving server. 

I tend to think of these particular tests as: if we replaced every `enqueueInitialChunk`/`enqueueSubsequentChunk` call with the real graphql.js `execute`, would the test work as expected with the only changes being the migration to the `execute` method. If not, then I think its a "bad" test (bad meaning not a real world example). Instead I'd prefer if we tested this using `cache.writeQuery` directly since that is where we'd see this. 

I'd move this to `cache.writeQuery/incremental.test.ts` to prove that we still see the missing field warnings for fields outside defer boundaries during writes.

File: src/core/__tests__/client.watchQuery/defer20220824.test.ts [hunk 18]
```diff
@@ -578,3 +579,170 @@ test("delivers cache updates written while a deferred response is still streamin
...
+    loading: false,
+    networkStatus: NetworkStatus.ready,
+    partial: false,
+  });
+
+  // this should not log errors, but currently does:
+  /*
+    1: "Missing field '%s' while writing result %o", "likes", {"__typename": "Comment", "author": {"__typename": "Author", "id": "a1", "name": "x"}, "id": "c1", "text": "first!"}
+    2: "Missing field '%s' while writing result %o", "badge", {"__typename": "Author", "id": "a1", "name": "x"}
+  */
```
Comment (jerelmiller) [inline src/core/__tests__/client.watchQuery/defer20220824.test.ts]: Is this an old comment? This test isn't marked as failing and tests seem to pass so I think it is?


```suggestion
```

**Review by jerelmiller — CHANGES_REQUESTED** [review]

File: src/incremental/handlers/defer20220824.ts (hunk 10 above)
Comment (jerelmiller) [reply src/incremental/handlers/defer20220824.ts]: Oh one more thing, I don't think the tests exercise this condition (`data: null`), at least from what I could see. Could you maybe add a test that verifies we don't see the warning when `this.hasFailedDefer` gets set?

File: src/cache/inmemory/writeToStore.ts (hunk 5 above)
Comment (phryneas) [reply src/cache/inmemory/writeToStore.ts:353]: cdd9557270364c7b21a9aca754edf3299fa37799

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts (hunk 9 above)
Comment (phryneas) [reply src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: 1adfe30268af8bdad692c748b1c38c879ca58ba7

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts (hunk 8 above)
Comment (phryneas) [reply src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: 1adfe30268af8bdad692c748b1c38c879ca58ba7

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts (hunk 6 above)
Comment (phryneas) [reply src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: 1adfe30268af8bdad692c748b1c38c879ca58ba7

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts (hunk 7 above)
Comment (phryneas) [reply src/cache/inmemory/__tests__/cache.diff/incremental.test.ts]: 1adfe30268af8bdad692c748b1c38c879ca58ba7

File: src/cache/inmemory/writeToStore.ts (hunk 4 above)
Comment (phryneas) [reply src/cache/inmemory/writeToStore.ts]: a60b5999b6beadba279bc1e10c109ded12553612

File: src/cache/inmemory/readFromStore.ts (hunk 3 above)
Comment (phryneas) [reply src/cache/inmemory/readFromStore.ts]: 5d4f589abebd3df6905b34cb1dcecfeed2682b2d

File: src/core/__tests__/client.watchQuery/defer20220824.test.ts (hunk 18 above)
Comment (phryneas) [reply src/core/__tests__/client.watchQuery/defer20220824.test.ts]: yeah that was a leftover e57aa5a542b1b9041e1228c77b2ecd79c94c5e7a

File: src/incremental/handlers/defer20220824.ts (hunk 12 above)
Comment (phryneas) [reply src/incremental/handlers/defer20220824.ts]: af847dcf8060e4572c14a8a71a742a513bd6d98d

File: src/incremental/handlers/defer20220824.ts (hunk 16 above)
Comment (phryneas) [reply src/incremental/handlers/defer20220824.ts]: af847dcf8060e4572c14a8a71a742a513bd6d98d

File: src/incremental/handlers/graphql17Alpha9.ts (hunk 15 above)
Comment (phryneas) [reply src/incremental/handlers/graphql17Alpha9.ts:134]: 813385abbd8138f8b4f9876a83442597a1610086

File: src/incremental/handlers/graphql17Alpha9.ts (hunk 13 above)
Comment (phryneas) [reply src/incremental/handlers/graphql17Alpha9.ts:148]: That would fully run `getPendingWithInfo` to create the full array and not be able to do an early exit though - it shouldn't be a performance concern, but it is slightly less performant.

Comment (phryneas) [reply src/incremental/handlers/graphql17Alpha9.ts:148]: What is your thought on having `getPendingWithInfo` return an iterator, and then using
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/map
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/some
(with relevant polyfills until the first minor after March 2027)

File: src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts (hunk 17 above)
Comment (phryneas) [reply src/core/__tests__/client.watchQuery/deferGraphQL17Alpha9.test.ts:11074]: I'm a bit torn here:

Yes, it fits better in a direct `writeQuery` test, but then, the "`writeQuery` with incremental" code path can only ever be reached from our own incremental handlers, not from a direct userland call.

If we were to argue that we only assume "correct" servers here, we could also go for a big shortcut and disable these warnings altogether for any data written by an incremental path - it's impossible that we'd ever get incomplete data, so why even waste the CPU time to test if it's complete?
